stepflow-rs/crates/stepflow-protocol/src/http/client.rs

```diff
@@ -11,315 +11,537 @@
             client,
-            outgoing_tx,
-            pending_tx,
-            loop_handle,
-        })
+            url,
+            context,
+            request_headers,
+        };
+
+        Ok(Self { handle })
     }
 
     pub fn handle(&self) -> HttpClientHandle {
-        HttpClientHandle {
-            client: self.client.clone(),
-            url: self.url.clone(),
-            outgoing_tx: self.outgoing_tx.clone(),
-            pending_tx: self.pending_tx.clone(),
-        }
+        self.handle.clone()
     }
 }
 
 #[derive(Clone)]
 pub struct HttpClientHandle {
     client: reqwest::Client,
     url: String,
-    outgoing_tx: mpsc::Sender<String>,
-    pending_tx: mpsc::Sender<(RequestId, oneshot::Sender<OwnedJson>)>,
+    context: Arc<dyn Context>,
+    request_headers: reqwest::header::HeaderMap,
 }
 
 impl HttpClientHandle {
-    pub async fn send_message(&self, message: String) -> Result<()> {
-        self.outgoing_tx
-            .send(message)
-            .await
-            .map_err(|_| TransportError::Send)?;
-        Ok(())
+    pub fn url(&self) -> &str {
+        &self.url
     }
 
-    pub async fn register_pending(
+    /// Send a typed method request and return the typed response
+    /// This follows the same pattern as StdioClientHandle
+    pub async fn method<I>(&self, params: &I) -> Result<I::Response>
+    where
+        I: ProtocolMethod + serde::Serialize + Send + Sync + std::fmt::Debug,
+        I::Response: DeserializeOwned + Send + Sync + 'static,
+    {
+        let response = self
+            .method_dyn(I::METHOD_NAME, LazyValue::write_ref(params))
+            .await?;
+
+        response
+            .value()
+            .deserialize_to()
+            .change_context(TransportError::InvalidResponse(I::METHOD_NAME))
+    }
+
+    async fn method_dyn(
         &self,
-        id: RequestId,
-        sender: oneshot::Sender<OwnedJson>,
-    ) -> Result<()> {
-        self.pending_tx
-            .send((id, sender))
+        method: Method,
+        params: LazyValue<'_>,
+    ) -> Result<OwnedJson<LazyValue<'static>>> {
+        let id = RequestId::new_uuid();
+        let request = MethodRequest {
+            jsonrpc: Default::default(),
+            id: id.clone(),
+            method,
+            params: Some(params),
+        };
+
+        let request_str = serde_json::to_string(&request)
+            .change_context(TransportError::SerializeRequest(method))?;
+
+        let response = self
+            .send_method_request(id.clone(), request_str)
+            .await?
+            .owned_response()?;
+
+        // This is an assertion since the routing should only send the response for the
+        // registered ID to the pending one-shot channel.
+        debug_assert_eq!(response.response().id(), &id);
+        response.into_success_value()
+    }
+
+    /// Send a typed notification (fire-and-forget)
+    pub async fn notify<I>(&self, params: &I) -> Result<()>
```

Comment:
Add a private `dyn` version with most of the code, so we don't need the many generic instantiations.